sim/rogue/fan_of_knives.go

```diff
@@ -0,0 +1,95 @@
+package rogue
+
+import (
+	"time"
+
+	"github.com/wowsims/cata/sim/core"
+	"github.com/wowsims/cata/sim/core/proto"
+)
+
+func (rogue *Rogue) registerFanOfKnives() {
+	fokSpell := rogue.RegisterSpell(core.SpellConfig{
+		ActionID:    core.ActionID{SpellID: 51723},
+		SpellSchool: core.SpellSchoolPhysical,
+		ProcMask:    core.ProcMaskRangedSpecial,
+		Flags:       core.SpellFlagMeleeMetrics | SpellFlagColdBlooded,
+
+		DamageMultiplier: 0.8 * (1 +
+			core.TernaryFloat64(rogue.Spec == proto.Spec_SpecCombatRogue, 0.75, 0.0)),
+		CritMultiplier:   rogue.MeleeCritMultiplier(false), // TODO (TheBackstabi, 3/16/2024) - Verify what crit table FoK is on
+		ThreatMultiplier: 1,
+	})
+
+	results := make([]*core.SpellResult, len(rogue.Env.Encounter.TargetUnits))
+
+	rogue.FanOfKnives = rogue.RegisterSpell(core.SpellConfig{
+		ActionID:    core.ActionID{SpellID: 51723},
+		SpellSchool: core.SpellSchoolPhysical,
+		Flags:       core.SpellFlagMeleeMetrics | core.SpellFlagAPL,
+
+		EnergyCost: core.EnergyCostOptions{
+			Cost: 35,
+		},
+		Cast: core.CastConfig{
+			DefaultCast: core.Cast{
+				GCD: time.Second,
+			},
+			IgnoreHaste: true,
+		},
+		ExtraCastCondition: func(sim *core.Simulation, target *core.Unit) bool {
+			return rogue.HasThrown()
+		},
+
+		ApplyEffects: func(sim *core.Simulation, unit *core.Unit, spell *core.Spell) {
+			rogue.BreakStealth(sim)
+			for i, aoeTarget := range sim.Encounter.TargetUnits {
+				baseDamage := fokSpell.Unit.RangedWeaponDamage(sim, fokSpell.RangedAttackPower(aoeTarget))
+				baseDamage *= sim.Encounter.AOECapMultiplier()
+
+				results[i] = fokSpell.CalcDamage(sim, aoeTarget, baseDamage, fokSpell.OutcomeRangedHitAndCrit)
+			}
+			for i, aoeTarget := range sim.Encounter.TargetUnits {
+				fokSpell.DealDamage(sim, results[i])
+
+				if rogue.Talents.VilePoisons > 0 {
+					poisonProcModifier := 0.33 * float64(rogue.Talents.VilePoisons)
+					mhProcChance := poisonProcModifier * getPoisonsProcChance(core.ProcMaskMeleeMH, rogue.Options.MhImbue, rogue)
+					ohProcChance := poisonProcModifier * getPoisonsProcChance(core.ProcMaskMeleeOH, rogue.Options.OhImbue, rogue)
```

Comment:
can this be calculated and cached outside of the ApplyEffects?

Reply:
poisonProcModifier can, but mh/ohProcChance being cached would remove the ability to increase poison application rate via Envenom.

That's about the only use-case I can think of, which is a very specific one but is a very strong interaction wherever applicable, but maybe not important in the context of a websim?